promql/generated_parser.y

```diff
@@ -35,120 +36,351 @@ import (
 
-%token	operatorsStart
 // Operators.
-%token <item> SUB
-%token <item> ADD
-%token <item> MUL
-%token <item> MOD
-%token <item> DIV
-%token <item> LAND
-%token <item> LOR
-%token <item> LUNLESS
-%token <item> EQL
-%token <item> NEQ
-%token <item> LTE
-%token <item> LSS
-%token <item> GTE
-%token <item> GTR
-%token <item> EQL_REGEX
-%token <item> NEQ_REGEX
-%token <item> POW
+%token	operatorsStart
+%token <item> ADD DIV EQL EQL_REGEX GTE GTR LAND LOR LSS LTE LUNLESS MOD MUL NEQ NEQ_REGEX POW SUB
 %token	operatorsEnd
 
-%token	aggregatorsStart
 // Aggregators.
-%token <item> AVG
-%token <item> COUNT
-%token <item> SUM
-%token <item> MIN
-%token <item> MAX
-%token <item> STDDEV
-%token <item> STDVAR
-%token <item> TOPK
-%token <item> BOTTOMK
-%token <item> COUNT_VALUES
-%token <item> QUANTILE
+%token	aggregatorsStart
+%token <item> AVG BOTTOMK COUNT COUNT_VALUES MAX MIN QUANTILE STDDEV STDVAR SUM TOPK
 %token	aggregatorsEnd
 
-%token	keywordsStart
 // Keywords.
-%token <item> OFFSET
-%token <item> BY
-%token <item> WITHOUT
-%token <item> ON
-%token <item> IGNORING
-%token <item> GROUP_LEFT
-%token <item> GROUP_RIGHT
-%token <item> BOOL
-
+%token	keywordsStart
+%token <item> BOOL BY GROUP_LEFT GROUP_RIGHT IGNORING OFFSET ON WITHOUT
 %token keywordsEnd
 
 
-%token	startSymbolsStart
 // Start symbols for the generated parser.
-%token START_LABELS
-%token START_METRIC
-%token START_GROUPING_LABELS
-%token START_SERIES_DESCRIPTION
+%token	startSymbolsStart
+%token START_METRIC START_SERIES_DESCRIPTION START_EXPRESSION START_METRIC_SELECTOR
 %token	startSymbolsEnd
 
-%type <matchers> label_matchers label_match_list
+
+// Type definitions for grammar rules.
+%type <matchers> label_match_list label_matchers
 %type <matcher> label_matcher
 
-%type <item> match_op metric_identifier grouping_label maybe_label
+%type <item> aggregate_op grouping_label match_op maybe_label metric_identifier unary_op
 
-%type <labels> label_set_list label_set metric
-%type <label> label_set_item    
-%type <strings> grouping_labels  grouping_label_list
-%type <series> series_values series_item
+%type <labels> label_set label_set_list metric
+%type <label> label_set_item
+%type <strings> grouping_label_list grouping_labels maybe_grouping_labels
+%type <series> series_item series_values
 %type <uint> uint
-%type <float> series_value signed_number number
+%type <float> number series_value signed_number
+%type <node> aggregate_expr aggregate_modifier bin_modifier binary_expr bool_modifier expr function_call function_call_args function_call_body group_modifiers matrix_selector number_literal offset_expr on_or_ignoring paren_expr string_literal subquery_expr unary_expr vector_selector
+%type <string> string
+%type <duration> duration maybe_duration
 
 %start start
 
+// Operators are listed with increasing precedence.
+%left LOR
+%left LAND LUNLESS
+%left EQL GTE GTR LSS LTE NEQ
+%left ADD SUB
+%left MUL DIV MOD
+%right POW
+
+// Offset modifiers do not have associativity.
```

Comment:
Is there an expression where this even comes up?

Reply:
The point this statement is to instruct yacc to evaluate expressions like `a + b offset 42s` as `a + (b offset 42s)` instead of `(a + b) offset 42s`. The latter would fail at a later stage anyway, but yacc does not know this.

Reply:
That's distributivity, not associativity.

Reply:
The `%nonassoc` here contains more than only associativity information:
 
The important part is that it is listed after the binary operators, which tells yacc that `offset` modifiers have a higher precedence than binary operators.

Every precedence specification in yacc must also specify associativity. Since the concept of associativity does not apply here`%nonassoc` is used.